torch/onnx/_type_utils.py

```diff
@@ -108,36 +113,86 @@ def from_name(
     @classmethod
     @_beartype.beartype
     def from_dtype(cls, dtype: torch.dtype) -> JitScalarType:
-        """Convert a torch dtype to ScalarType."""
+        """Convert a torch dtype to JitScalarType.
+
+        NB: DO NOT USE this API if dtype is comes from a torch._C.Value.type().dtype() call because
+            "RuntimeError: INTERNAL ASSERT FAILED at "../aten/src/ATen/core/jit_type_base.h" can
+            be raised in some scenarios. Instead use from_value API which handles all cases
+
+        Args:
+            dtype: A torch.dtype to create a JitScalarType from
+
+        Returns:
+            JitScalarType
+
+        Raises:
+            ValueError: if dtype is not a valid torch.dtype or if it is None.
+        """
         if dtype not in _DTYPE_TO_SCALAR_TYPE:
             raise ValueError(f"Unknown dtype: {dtype}")
         return _DTYPE_TO_SCALAR_TYPE[dtype]
 
+    # TODO: `value: torch._C.Value` type annotation raises (CI machines only):
+    #       NameError: name '_C' is not defined
+    # In both cases, using typing.TypeVar or string literals didn't work
+    @classmethod
+    @_beartype.beartype
+    def from_value(cls, value) -> JitScalarType:
+        """Create a JitScalarType from a torch._C.Value underlying scalar type.
+
+        Args:
+            value: A `torch._C.Value` object to fetch scalar type from.
+
+        Returns:
+            JitScalarType.
+
+        Raises:
+            OnnxExporterError: if value is None.
+            SymbolicValueError: when value.type().scalarType() does not exist
+
+        """
+
+        if value is None:
+            raise errors.OnnxExporterError(
+                "Cannot determine scalar type for `None` instance."
+            )
+        elif isinstance(value, torch.Tensor):
+            return JitScalarType.from_dtype(value.dtype)
+        elif isinstance(value.type(), torch.ListType):
+            return JitScalarType.from_dtype(value.type().getElementType().dtype())
+        scalar_type = value.type().scalarType()
+        if scalar_type is None:
+            raise errors.SymbolicValueError(
```

Comment:
A ValueError seems more suitable semantically, although I can see the value of having a context for `value` in the error message 🤔

Reply:
Yeah, the goal was to expose `value` into the error message

On a side note, https://github.com/pytorch/pytorch/issues/76155 suggests (using `RuntimeError` as a real use case, but `ValueError` follows the same idea), we don't want to use pure python exceptions inside `torch.onnx`  to be able to distil between exporter intentionally raised exceptions and everything else (custom ops exceptions, torchlib backend exceptions, etc), allowing aten fallback and possibly other features/user extensions to handle exceptions unequivocally.

https://github.com/pytorch/pytorch/pull/78446/files#diff-db432e52aff4342354a4c07eef78e2d26966605c219e5e2635727bb99c9e9824R10 almost did that, as it implemented the concept of abstracting the exporter-specific exceptions, but it killed the principle by inheriting from `RuntimeError`. The BC in this case was not desired. Hopefully we can refactor it soon to address the [issue](https://github.com/pytorch/pytorch/issues/76155)